kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs

```diff
@@ -304,24 +304,58 @@ impl<'tcx> GotocCtx<'tcx> {
             }
             (Scalar::Int(_), ty::Adt(adt, subst)) => {
                 if adt.is_struct() || adt.is_union() {
-                    // in this case, we must have a one variant ADT. there are two cases
-                    let variant = &adt.variants().raw[0];
-                    // if there is no field, then it's just a ZST
+                    // In this case, we must have a one variant ADT. There are two cases.
+                    let variant = adt.non_enum_variant();
+                    let overall_type = self.codegen_ty(ty);
+                    // If there is no field, then it's just a ZST
                     if variant.fields.is_empty() {
                         if adt.is_struct() {
-                            let overall_t = self.codegen_ty(ty);
-                            Expr::struct_expr_from_values(overall_t, vec![], &self.symbol_table)
+                            Expr::struct_expr_from_values(overall_type, vec![], &self.symbol_table)
                         } else {
                             unimplemented!()
                         }
                     } else {
-                        // otherwise, there is just one field, which is stored as the scalar data
-                        let field = &variant.fields[0usize.into()];
-                        let fty = field.ty(self.tcx, subst);
-
-                        let overall_t = self.codegen_ty(ty);
+                        // Otherwise, there is at least one field. Again, there are two cases.
+                        let overall_type = self.codegen_ty(ty);
                         if adt.is_struct() {
-                            self.codegen_single_variant_single_field(s, span, overall_t, fty)
+                            if variant.fields.len() == 1 {
+                                // There is a single field which we associate with the scalar data.
+                                let field = variant.single_field();
+                                let field_type = field.ty(self.tcx, subst);
+                                self.codegen_single_variant_single_field(
+                                    s,
+                                    span,
+                                    overall_type,
+                                    field_type,
+                                )
+                            } else {
+                                // There are multiple fields, but only one is related to the scalar data.
+                                // The rest of them correspond to phantom data (ZST).
+                                let field_types: Vec<Ty<'_>> =
+                                    variant.fields.iter().map(|f| f.ty(self.tcx, subst)).collect();
+                                // Check that there is a single non-phantom field
```

Comment:
This is not only phantom data related. Any ZST can trigger this issue. For example:

```rust
struct Empty {}

pub struct Foo {
    x: u8,
    _t: Empty,
}

#[kani::proof]
fn check_zst() {
    const C: Foo = Foo { x: 0, _t: Empty {} };
    assert_eq!(C.x, 0);
}
```
So please update the comments and variables to reflect that.

Reply:
True. I believe all comments, variables and tests have been updated, please take a look.